File: frontend/types/transformation.ts
/**
 * SQL Transformation Types
 */

export type TransformationStatus = "pending" | "running" | "completed" | "failed" | "skipped";

export type WriteMode = "replace" | "append" | "merge";

export interface SQLTransformation {
  id: string;
  pipeline_id: string;
  name: string;
  description?: string;
  sql_query: string;
  target_table?: string;
  write_mode: WriteMode;
  execution_order: number;
  is_active: boolean;
  continue_on_error: boolean;
  timeout_seconds: number;
  created_at: string;
  updated_at: string;
}

export interface CreateTransformationRequest {
  name: string;
  description?: string;
  sql_query: string;
  target_table?: string;
  write_mode?: WriteMode;
  execution_order?: number;
  is_active?: boolean;
  continue_on_error?: boolean;
  timeout_seconds?: number;
}

export interface UpdateTransformationRequest {
  name?: string;
  description?: string;
  sql_query?: string;
  target_table?: string;
  write_mode?: WriteMode;
  execution_order?: number;
  is_active?: boolean;
  continue_on_error?: boolean;
  timeout_seconds?: number;
}

export interface ReorderTransformationsRequest {
  transformation_ids: string[];
}

export interface TransformationResult {
  id: string;
  pipeline_run_id: string;
  transformation_id: string;
  status: TransformationStatus;
  started_at?: string;
  completed_at?: string;
  duration_seconds?: number;
  rows_affected?: number;
  error_message?: string;
  error_traceback?: string;
  result_metadata?: Record<string, any>;
  created_at: string;
}

export interface SQLPreviewRequest {
  sql_query: string;
  limit?: number;
}

export interface SQLPreviewResult {
  columns: string[];
  rows: Record<string, any>[];
  row_count: number;
  execution_time_ms: number;
  error?: string;
  error_traceback?: string;
}

File: backend/models/transformation.py
"""
SQL Transformation ORM Models.

Defines database models for SQL transformations that run after
data is loaded to the destination.
"""
from datetime import datetime, timezone
from typing import Optional
import uuid
import enum

from sqlalchemy import (
    Column,
    Integer,
    String,
    DateTime,
    Boolean,
    Text,
    ForeignKey,
    JSON,
    Enum as SQLEnum,
    Float,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from backend.database import Base


class TransformationStatus(str, enum.Enum):
    """Transformation execution status."""
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    SKIPPED = "skipped"


class SQLTransformation(Base):
    """
    SQL Transformation model.

    Represents a SQL transformation that runs against the destination
    after data is loaded. Multiple transformations can be defined per
    pipeline and they execute in order.
    """
    __tablename__ = "sql_transformations"

    id = Column(Integer, primary_key=True, index=True)
    public_id = Column(UUID(as_uuid=True), unique=True, nullable=False, default=uuid.uuid4, index=True)
    pipeline_id = Column(Integer, ForeignKey("pipelines.id", ondelete="CASCADE"), nullable=False, index=True)

    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)

    # SQL query to execute
    sql_query = Column(Text, nullable=False)

    # Target table for results (optional - if not set, SQL runs as statement)
    target_table = Column(String(255), nullable=True)

    # Write mode: replace, append, merge
    write_mode = Column(String(50), nullable=True, default="replace")

    # Execution order (lower numbers execute first)
    execution_order = Column(Integer, nullable=False, default=0)

    # Whether this transformation is active
    is_active = Column(Boolean, default=True, nullable=False)

    # Continue on error (if True, pipeline continues even if this transform fails)
    continue_on_error = Column(Boolean, default=False, nullable=False)

    # Timeout in seconds (0 = no timeout)
    timeout_seconds = Column(Integer, nullable=False, default=300)

    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc), nullable=False)

    # Relationships
    pipeline = relationship("Pipeline", back_populates="sql_transformations")
    execution_results = relationship("TransformationResult", back_populates="transformation", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<SQLTransformation(id={self.id}, name='{self.name}', order={self.execution_order})>"


class TransformationResult(Base):
    """
    Transformation Result model.

    Records the result of executing a transformation during a pipeline run.
    """
    __tablename__ = "transformation_results"

    id = Column(Integer, primary_key=True, index=True)
    public_id = Column(UUID(as_uuid=True), unique=True, nullable=False, default=uuid.uuid4, index=True)
    pipeline_run_id = Column(Integer, ForeignKey("pipeline_runs.id", ondelete="CASCADE"), nullable=False, index=True)
    transformation_id = Column(Integer, ForeignKey("sql_transformations.id", ondelete="CASCADE"), nullable=False, index=True)

    status = Column(SQLEnum(TransformationStatus), default=TransformationStatus.PENDING, nullable=False)

    # Execution timing
    started_at = Column(DateTime, nullable=True)
    completed_at = Column(DateTime, nullable=True)
    duration_seconds = Column(Float, nullable=True)

    # Result metrics
    rows_affected = Column(Integer, nullable=True)

    # Error information
    error_message = Column(Text, nullable=True)
    error_traceback = Column(Text, nullable=True)

    # Additional metadata (query plan, etc.)
    result_metadata = Column(JSON, nullable=True)

    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)

    # Relationships
    pipeline_run = relationship("PipelineRun", back_populates="transformation_results")
    transformation = relationship("SQLTransformation", back_populates="execution_results")

    def __repr__(self):
        return f"<TransformationResult(id={self.id}, transformation_id={self.transformation_id}, status='{self.status}')>"
